### PR TITLE
Fixed inability to build models

### PR DESCRIPTION
**Pull request by ilia-nikiforov-umn**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This PR adds the ability for the KIM API conda-forge package to build KIM models from source. This involves adding several compilers and build tools as run requirements and passing their paths to CMake to configure the KIM API. I also added a test for building a model.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -26,10 +26,16 @@ requirements:
...
+    - cmake
+    - make
+    - {{ compiler('cxx') }}
+    - {{ compiler('c') }}
+    - {{ compiler('fortran') }}
 
 test:
   commands:
     - kim-api-collections-management list
+    - kim-api-collections-management install CWD SW_StillingerWeber_1985_Si__MO_405512056662_006
```
Comment (jan-janssen) [inline recipe/meta.yaml]:
```suggestion
    - kim-api-collections-management install CWD SW_StillingerWeber_1985_Si__MO_405512056662_006  # [linux]
```

Comment (jan-janssen) [reply recipe/meta.yaml]: By adding `# [linux]` at the end of the line the test is only executed on linux, as the model is not supported on Mac Os as far as I understand.

Comment (ilia-nikiforov-umn) [reply recipe/meta.yaml]: @jan-janssen It is working on osx_arm64 according to the CI checks and I hope we can get it working on osx_64 with a small tweak

edit: Nevermind, I guess the osx_arm64 doesn't run any tests? Nevertheless, I hope we can figure out building on osx_64

**Comment by mjwen** [comment]
To get it to work on osx_64, instead of using 
```
    - {{ compiler('cxx') }}
    - {{ compiler('c') }}
    - {{ compiler('fortran') }}
 ```
which I guess finds the default compilers on the system, is it possible specify some settings such that it installs other compilers (e.g. gcc suits) and then uses those? 

I completely have no idea of how conda-forge works. So the above guess might make no sense.

**Comment by ellio167** [comment]
I think we’ll need to figure out how this is all meant to be used at run time. I would hope conda has a coherent rational for its environment variable settings across architectures…. It’ll probably require some further digging and comparison of how existing packages are dealing with this issue…RyanOn Dec 6, 2023, at 9:20 AM, Mingjian Wen ***@***.***> wrote:﻿
To get it to work on osx_64, instead of using
    - {{ compiler('cxx') }}
    - {{ compiler('c') }}
    - {{ compiler('fortran') }}

which I guess finds the default compilers on the system, is it possible to use some other settings such that it installs other compilers (e.g. gcc suits) and use it?
I completely have no idea of how conda-forge works. So the above guess might make no sense.

**Comment by ilia-nikiforov-umn** [comment]
> I think we’ll need to figure out how this is all meant to be used at run time. I would hope conda has a coherent rational for its environment variable settings across architectures…. It’ll probably require some further digging and comparison of how existing packages are dealing with this issue…

I think all we need to do is figure out how to get CMake to build the KIM API with `KIM_API_CMAKE_C_COMPILER` set to `$ENV{CC}`, and analogously for the other compilers. The difficulty is that, if I try the following configuration option:

`-DKIM_API_CMAKE_C_COMPILER="\$ENV{CC}"`, then CMake thinks it is a relative path (because the variable type for the cache variable `KIM_API_CMAKE_C_COMPILER` is `FILEPATH`). So I add a leading forward slash, `-DKIM_API_CMAKE_C_COMPILER="/\$ENV{CC}"`, to get CMake to accept it as an absolute path. This works on Linux, because `$ENV{CC}=/home/ilia/miniconda3/envs/kim-env-pr/bin/x86_64-conda-linux-gnu-cc` or whatever, so an additional leading forward slash doesn't break anything.


On OSX (I'm getting this from the CI logs, since I don't have a Mac, but it seems pretty straightforward), `$ENV{CC}=x86_64-apple-darwin13.4.0-clang`, so the leading slash makes it look for `/x86_64-apple-darwin13.4.0-clang`, which is not a valid path. I assume the conda-forge compiler package adds the wherever `x86_64-apple-darwin13.4.0-clang` is to `$PATH`, so if we can just get CMake to set  `KIM_API_CMAKE_C_COMPILER=$ENV{CC}` instead of `KIM_API_CMAKE_C_COMPILER=/$ENV{CC}` as it stands now, everything would work.

**Comment by mjwen** [comment]
I've taken a look at the CI log, and seems we can have a solution for it. 

On linux_64_, the CC and FC environment variables are indeed a full path (image below).  
Screenshot 2023-12-06 at 5 12 14 PM

But on osx_64_, while FC/F90 is full path, CC is not (image below). I guess this is the reason why @ilia-nikiforov-umn sees the error.  
Screenshot 2023-12-06 at 5 15 00 PM

But, as you might have noticed in the images, CC is essentially the same as CC_FOR_BUILD. Similarly, FC is the same as FC_FOR_BUILD. The nice thing is that CC_FOR_BUILD and FC_FOR_BUILD are full paths. 

So, I believe we can modify `build.sh` to use **_FOR_BUILD. 

We can even use the GCC suits by choosing the corresponding environment variable (but I don't think that's a good choice).

**Comment by ilia-nikiforov-umn** [comment]
I don't think that will work for platforms that are cross-compiled, like
osx_arm64. See the build log here:
https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/837612/logs/24

CC_FOR_BUILD and CC are different compilers here.

**Comment by ellio167** [comment]
Cmake has some tools to find things (files, executables, … I think). So maybe we could use one of those with the value if the environment variable to turn it into a path?Ryan Sent from my iPhoneSEhttps://z.umn.edu/relliottOpenPGP: https://z.umn.edu/relliott_asc (576D4997C4D51D92)On Dec 6, 2023, at 3:02 PM, ilia-nikiforov-umn ***@***.***> wrote:﻿

I think we’ll need to figure out how this is all meant to be used at run time. I would hope conda has a coherent rational for its environment variable settings across architectures…. It’ll probably require some further digging and comparison of how existing packages are dealing with this issue…

I think all we need to do is figure out how to get CMake to build the KIM API with KIM_API_CMAKE_C_COMPILER set to $ENV{CC}, and analogously for the other compilers. The difficulty is that, if I try the following configuration option:
-DKIM_API_CMAKE_C_COMPILER="\$ENV{CC}", then CMake thinks it is a relative path (because the variable type for the cache variable KIM_API_CMAKE_C_COMPILER is FILEPATH). So I add a leading forward slash, -DKIM_API_CMAKE_C_COMPILER="/\$ENV{CC}", to get CMake to accept it as an absolute path. This works on Linux, because $ENV{CC}=/home/ilia/miniconda3/envs/kim-env-pr/bin/x86_64-conda-linux-gnu-cc or whatever, so an additional leading forward slash doesn't break anything.
On OSX (I'm getting this from the CI logs, since I don't have a Mac, but it seems pretty straightforward), $ENV{CC}=x86_64-apple-darwin13.4.0-clang, so the leading slash makes it look for /x86_64-apple-darwin13.4.0-clang, which is not a valid path. I assume the conda-forge compiler package adds the wherever x86_64-apple-darwin13.4.0-clang is to $PATH, so if we can just get CMake to set  KIM_API_CMAKE_C_COMPILER=$ENV{CC} instead of KIM_API_CMAKE_C_COMPILER=/$ENV{CC} as it stands now, everything would work.

**Comment by mjwen** [comment]
> I don't think that will work for platforms that are cross-compiled, like osx_arm64. See the build log here: https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/837612/logs/24 CC_FOR_BUILD and CC are different compilers here.

OK, I guess we will have to use CC, FC, and such then. 

Is there any chance that CC in osx_64_ is incorrectly set by conda-forge? I meant, now it is `+CC=×86_64-apple-darwin13.4.0-clang`, but it should be something like `+CC=$BUILD_PREFIX/bin/×86_64-apple-darwin13.4.0-clang`? 

It this is the case and conda-forge can fix that. We are good. @jan-janssen any idea?

**Comment by jan-janssen** [comment]
I presume it is related to cross-compiling on Apple systems. Like naively I thought you could try something like: 
```
export CC=$BUILD_PREFIX/bin/$CC
```
Still looking more closely, at the Apple ARM log files: 
```
CC=arm64-apple-darwin20.0.0-clang
CC_FOR_BUILD=$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang
```
So it seems the `arm64-apple-darwin20.0.0-clang` is located somewhere else. You could check the full path of the compiler using something like `which`. In the error message the compiler is listed as: 
```
CMake Error at CMakeLists.txt:40 (project):
  The CMAKE_CXX_COMPILER:

    /x86_64-apple-darwin13.4.0-clang++

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.


CMake Error at CMakeLists.txt:40 (project):
  The CMAKE_C_COMPILER:

    /x86_64-apple-darwin13.4.0-clang

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.
```
I feel the issue is the `/` in front of the compiler name. Somehow it expected to save the absolute path but then only stored the variable plus the `/`.

**Comment by ilia-nikiforov-umn** [comment]
It appears that specifying file type for the compiler paths on the command line (i.e. `-D<var>:<type>=<value>` instead of `-D<var>=<value>`) prevents CMake from messing with `CACHE` entries of type `FILEPATH`, allowing me to set the paths to just `$ENV{CC}` etc. like I wanted.

I am not sure if this is intended -- the documentation is unclear: https://cmake.org/cmake/help/latest/manual/cmake.1.html#generate-a-project-buildsystem:

> If the `:<type>` portion is given it must be one of the types specified by the [set()](https://cmake.org/cmake/help/latest/command/set.html#command:set) command documentation for its `CACHE` signature. If the `:<type>` portion is omitted the entry will be created with no type if it does not exist with a type already. If a command in the project sets the type to `PATH` or `FILEPATH` then the `<value>` will be converted to an absolute path.

To me, this is ambiguous whether the last sentence is meant to apply always, or only when `:<type>` is omitted, but the functionality appears to be the latter. Maybe I'll ask on the CMake forum if this is intended.

Regardless, I think this is ready to merge, I don't think this will break anything that was already working, and it's passing checks including building the model on all platforms except osx_arm64, which does not run the checks (does anyone know why?)

Maybe someone with an Apple Silicon Mac wants to test it locally?

**Comment by ilia-nikiforov-umn** [comment]
I think the osx_arm64 build doesn't run checks because it's being cross-compiled, so whatever cloud resource is building it can't actually run it.

@jan-janssen I think this is ready to merge.

**Review by ClaireWaters — APPROVED** [review]
I'm not an expert on conda, but I don't see anything likely to break the pipeline side of things. If everyone is satisfied I'm ready to merge.

**Review by jan-janssen — APPROVED** [review]
Looks good to me as well

**Pull request merged**